### ISSUE TITLE
Allow urllib3 2.0.0 to be installed

### ISSUE THREAD

**Issue by epicserve** [boto/botocore#2926]
### Describe the feature

There is a new version of urllib3 (v2.0.0) that can't be installed in projects that use botocore because, in the setup.py file, it has the version restriction `urllib3>=1.25.4,<1.27`.

### Use Case

So other projects that want to upgrade to urllib3 2.0.0 can without being blocked by the botocore library.

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### SDK version used

latest

### Environment details (OS name and version, etc.)

all

**Comment by nateprewitt** [comment]
Hi @epicserve, thanks for reaching out. The pin is currently intentional while we wait for the major version bump to stabilize. We intend to move the pin shortly after Requests does to ensure we're not negatively impacting users.

We'll keep this issue updated and feel free to follow along in psf/requests#6432 as well if you like.

**Comment by jplock** [comment]
I started getting this error message in my Lambda logs after doing a recent deploy which I believe is pulling in urllib3 >=2.0.0

```
Runtime.ImportModuleError: Unable to import module 'app.lambda_handler': cannot import name 'DEFAULT_CIPHERS' from 'urllib3.util.ssl_' (/opt/python/urllib3/util/ssl_.py)
```

which is coming from https://github.com/boto/botocore/blob/master/botocore/httpsession.py#L52

```python
try:
    from urllib3.util.ssl_ import DEFAULT_CIPHERS
except ImportError:
    # Defer to system configuration starting with
    # urllib3 2.0. This will choose the ciphers provided by
    # Openssl 1.1.1+ or secure system defaults.
    DEFAULT_CIPHERS = None
```

which is concerning because urllib3 removed `DEFAULT_CIPHERS` in https://github.com/urllib3/urllib3/pull/2705 which is part of v2.0.0

What I'm not understanding is why the `ImportError` isn't being caught?

**Comment by nateprewitt** [comment]
Hi @jplock, this was already resolved in https://github.com/boto/botocore/pull/2924. You'll likely want to ensure you're not upgrading to urllib3 2.0 for the time being or add a more recent version of Botocore to your Lambda as detailed [here](https://docs.aws.amazon.com/lambda/latest/dg/python-package.html).

**Comment by pistolilla** [comment]
Same as @jplock here. I started seeing this error today on a lambda after I made deployment. Tried rolling back but error remained.

**Comment by nateprewitt** [comment]
Hi @jplock, if you're still encountering this error you'll want to add this to your dependency closure in the same way you're specifying the dependencies you've listed above. Make sure this is the last entry:

```
urllib3<2
```

One of your dependencies is pulling in the latest version of urllib3 which contains a handful of breaking changes.

**Comment by edgarrmondragon** [comment]
`requests` 2.30 added support for urllib3 2.0: https://github.com/psf/requests/releases/tag/v2.30.0

**Comment by mgorny** [comment]
As of requests-2.30.0, urllib3-2.0* is already allowed. This package is now blocking urllib3-2.0* upgrade for Gentoo users.

**Comment by nateprewitt** [comment]
Hi @mgorny, we're aware that Requests has moved their pin. We've seen there are a number of breaking cases in common workflows for AWS customers with the introduction of the new major version which has slowed how quickly we might move towards using urllib3 2.0 by default. 

Botocore is properly pinned for version support at the moment to limit unnecessary blast radius. Gentoo is free to move the constraints for their distribution of botocore if there's a strong need for urllib3 2.0 migration. I don't think that can reasonably done here with the auto-upgrade behavior of PyPI though.

**Comment by mgorny** [comment]
Could you at least fix the test suite to pass with urllib3 2.0*, please? IIUC the mocks currently rely on some internal behavior that's no longer correct.

**Comment by vazome** [comment]
Same: `botocore 1.29.102 requires urllib3<1.27,>=1.25.4, but you have urllib3 2.0.2 which is incompatible.`

**Comment by ahvigil** [comment]
The conflict here with what a common package like `requests` has pinned, combined with the prolific release cadence of botocore, makes this especially inconvenient for projects using poetry. The resolver there is almost comically bad at traversing the many dead ends this produces in resolving the dependency graph.

https://github.com/python-poetry/poetry/issues/7858
https://github.com/python-poetry/poetry/issues/8074

**Comment by nateprewitt** [comment]
Hi @ahvigil, to quickly clarify, Requests _does not_ have urllib3 2.0 pinned. I provided [how you'd do this](https://github.com/psf/requests/issues/6443#issuecomment-1535667256) with poetry in the primary Requests thread.

**Comment by ahvigil** [comment]
> Hi @ahvigil, to quickly clarify, Requests _does not_ have urllib3 2.0 pinned. I provided [how you'd do this](https://github.com/psf/requests/issues/6443#issuecomment-1535667256) with poetry in the primary Requests thread.

yes thats the suggested workaround in the linked poetry issues as well. Doing so nudges poetry in the right direction, reduces the dependency search space, and everything works great. Its just an odd thing to have to do to avoid combinatorial explosions for two pretty commonly used packages. Obviously `botocore` devs aren't responsible for how poetry performs for this one weird edge case but just thought I'd make a note of it.

**Comment by tjni** [comment]
Hi team, please forgive me bumping this to draw attention to https://github.com/boto/botocore/pull/2984. This is my attempt to move us closer to being able to support urllib3>=2 by addressing the failing unit tests around the "100 Continue" code path (which are catching a legitimate issue). It ought to also allow urllib3 to stop skipping these tests in its CI since https://github.com/urllib3/urllib3/pull/2565 was merged.

Feedback appreciated! Hopefully it is of high enough quality to be iterated on and merged, or perhaps Amazon is exploring a different approach internally.

**Comment by ai-robert** [comment]
I can't upload anything to s3 from aws lambda because this isn't fixed.

**Comment by nateprewitt** [comment]
@mgorny, providing a quick update that the test suite should be compatible with urllib3 2.0 now. We removed the upstream patch in urllib3 with https://github.com/urllib3/urllib3/pull/3098. Please let me know if you're still seeing issues.

**Comment by mgorny** [comment]
Thank you. I can confirm that the tests now pass w/ botocore 1.31.15 and urllib3 2.0.4.

**Comment by utkonos** [comment]
Is there anything else that is needed, or can we upgrade to urllib3 v2?

**Comment by mgorny** [comment]
We've been allowing urllib3-2 in Gentoo since 2023-05-19 and we haven't had any reports of problems. That said, I have no clue how many people actually use these packages.

**Comment by mridul97-dev** [comment]
any update on this? Need to install a compatible version of botocore with urllib>2.0
@tjni @nateprewitt @ahvigil

**Comment by tjni** [comment]
Hi @mridul97-dev, to clarify: I'm working on packaging this in [nixpkgs](https://github.com/NixOS/nixpkgs), which builds and distributes Python packages from source independently of PyPI, which offers us the possibility to patch botocore to lift the version cap on urllib3. I believe that @mgorny is in a similar situation with Gentoo. We can do this in part because our part of the world is smaller (e.g. for me, just nixpkgs users).

If you are using stock poetry, you'll need to rely on the maintainers of this package (e.g. @nateprewitt) to lift the restriction when they feel it is ready for a wider audience.

**Comment by nateprewitt** [comment]
You can currently install urllib3 2.0.x with boto3/botocore/awscli and it will work as we've unblocked Gentoo and nixpkgs to do. The primary reason that the pin is still in place for general python installations is because urllib3 2.0 has compatibility issues with Amazon Linux 2 and some versions of AWS Lambda. These are detailed in the urllib3 migration guide [here](https://urllib3.readthedocs.io/en/stable/v2-migration-guide.html#ssl-module-is-compiled-with-openssl-1-0-2-k-fips).

Because of the way pip works by always choosing to install the latest version of packages, moving the pin for urllib3 will unexpectedly break everyone not pinning urllib3 or using OpenSSL 1.1.1+. That's something we take very seriously with AWS products and try to ensure never happens. We're still working on a plan to smoothly migrate this customer base, but don't have any immediate timeline.

For the time being, users are free to install urllib3 2.x _after_ they've installed boto3/botocore/awscli, and should be able to use each package. Something like this would work in automated environments returning a successful response code. We realize it's not ideal, but is the most tractable solution we have currently that maintains stability:

```
python -m pip install boto3==x.y.z
python -m pip install urllib3==2.0.4 --no-deps
```

**Comment by mridul97-dev** [comment]
Hi @nateprewitt, @tjni  thank-you for providing a workaround for this. Allow me to explain my usecase - currently we use poetry for all dependency management - we came upon a problem of `backoff_max` parameter which only gets added within a patch of urllib3=2.0.x, but updating this version within poetry is resulting in failures as botocore is complaining that it can only go as high as urllib<1.7. 
--
As an alternative workaround, i took out urllib3 from the poetry env and tried installing it separately with `pip install urllib3==2.0.4` but this resulted in another import error `cannot import name 'DEFAULT_CIPHERS' from 'urllib3.util.ssl_'`.
--

From what i can see, neither poetry OR poetry+pip seems to be agreeing with the library installation - which is worrysome at our end.

**Comment by nateprewitt** [comment]
> this resulted in another import error cannot import name 'DEFAULT_CIPHERS' from 'urllib3.util.ssl_'.

You may want to ensure you've updated Boto3 and Botocore in your poetry lock file before installing urllib3. This issue was patched in April with botocore 1.29.123.

**Comment by SamStephens** [comment]
This is going to start to become a real problem as libraries start to insist on urllib 2.x - [google-auth just has](https://github.com/googleapis/google-auth-library-python/blob/main/CHANGELOG.md#user-content-2231-2023-09-26) for example.

The hack you provide with `--no-deps` does not work in environments where you're installing requirements from requirements.txt.

We're going to end up stuck on old versions of dependencies, and someone will miss a security fix because of it.

**Comment by ThiefMaster** [comment]
wait, this STILL hasn't been fixex? :/

**Comment by chrisdlangton** [comment]
I'm just surprised they haven't vendored it to avoid version conflicts like this..
They did it with `requests` forever.

I realise this comment is not constructive 'to maintainers' because my aim was to set expectations here.
This project is dying and it's no person's fault, it's just the way Amazon treats Python (or rather avoids it) and the good folks trying to cover all the python needs need at least 100 more people to help them - i'm seeing this across all of AWS consulting with government, FSI, startups, banks, everyone feels this pain for at least 5 years and it get's longer each time.

We see current stable python version only make it into Lambda runtimes when 3 releases behind, for a while it was 2 (3.7 came when we had 3.9), recently we got 3.10 when we were 4 behind and now still behind, almost 2 versions behind, at the time it was released!

Don't get me started on pSpark (Glue EMR Sagemaker) or anything else, the Lambda was a good new story, we got 3.10, forgetting 3.12 is pretty much ready, we actually got 3.10 which is a 'W' - the other python bits across AWS are worse than this good (actually bad) news!

**Comment by jessemyers-lettuce** [comment]
FYI, `types-requests` is now requiring `urllib3>=2.0.0`; this fact greatly increases the likelihood that other HTTP-oriented projects will force a newer version and will be in conflict with `boto*`.

**Comment by nateprewitt** [comment]
Hi everyone, for any readers who didn't see [the previous post](https://github.com/boto/botocore/issues/2926#issuecomment-1673963526), lack of support for urllib3 2.0 in Amazon Linux 2 and some versions of Lambda are what is currently blocking us removing the pin. These compatibility issues are unlikely to change so we're continuing to find options minimize the current dependency friction.

I've put up #3034 which will relax our pin to allow urllib3 2.0 on Python 3.10 and later. This should allow users on versions of Python that only support OpenSSL 1.1.1+ to safely upgrade. We'll continue to keep this thread updated as we have more to share.

**Comment by nateprewitt** [comment]
Hi everyone, for any readers who didn't see [the previous post](https://github.com/boto/botocore/issues/2926#issuecomment-1673963526), lack of support for urllib3 2.0 in Amazon Linux 2 and some versions of Lambda are what is currently blocking us removing the pin. These compatibility issues are unlikely to change so we're continuing to explore options to minimize the dependency friction.

I've put up #3034 which will relax our pin to allow urllib3 2.0 on Python 3.10 and later. This should allow users on versions of Python that only support OpenSSL 1.1.1+ to safely upgrade. We'll continue to keep this thread updated as we have more to share.

**Comment by jessemyers-lettuce** [comment]
Thanks @nateprewitt, completely understand the reason why this is pinned and the challenges you face. I just wanted to share new context -- as of this week and led to a great discussion in `typeshed` --  in case it wasn't widely known; I was certainly caught by surprise.

**Comment by xcaptain** [comment]
@nateprewitt  I can now install urllib3 2.0 with boto3, but now my tests reported some warnings, after some debugging, I thought it's related to:
https://github.com/boto/botocore/blob/develop/botocore/httpsession.py#L338
![image](https://github.com/boto/botocore/assets/4054836/568a1023-7a9f-40cd-872c-7824fd193211)

Maybe you should consider remove the `strict` option?

**Comment by nateprewitt** [comment]
Thanks for the report, @xcaptain. We're currently pinned at urllib3<2.1 since it will contain a handful of breaking changes. We'll take a look at getting this addressed to reduce warning noise, but there shouldn't be any other impact from these. For the time being, you should be able to suppress the deprecations or downgrade to "urllib3<2" if needed while we get a patch together.

**Comment by nateprewitt** [comment]
@Morgomor could you provide more info on what import error occurred? We have reasonable adoption on 3.10 now without issue so it sounds like you may have a unique setup.

**Comment by brodybits** [comment]
Is there anything we can do to help unblock an upgrade to a more recent version of `urllib3`, especially for those using Python 3.11 or 3.12?

**Comment by nateprewitt** [comment]
Hi @brodybits, botocore currently supports `urllib3<2.1` for Python 3.10+. There are some breaking changes in 2.1 (and more recently 2.2.0) that have prompted the current pinning. Once we have confidence we can safely move to the latest release, we'll look at updating the pin to catch up to the last couple months of releases.

**Comment by nateprewitt** [comment]
Given we've resolved the 2.0 move on Python 3.10+ and don't have any immediate plans to add support for earlier versions, it's probably best to close this issue.

For tracking catching up to the latest minor release we can have a separate feature request, that will help with some of the confusion this thread has had in the last few months. Thanks everyone for your participation and patience!

**Comment by SamStephens** [comment]
@nateprewitt can you please give us a link to the feature request for the update to urllib >= 2.1?

**Comment by brodybits** [comment]
@SamStephens I just filed https://github.com/boto/boto3/issues/4025

**Comment by plexosoft** [comment]
Hello there,

GitHub Dependabot is recommending an upgrade to `urllib3>=2.5.0` due to a moderate security vulnerability. However, when I attempt the upgrade, I run into a dependency conflict with `botocore==1.31.63`, which is required for interacting with AWS S3.

Here’s the relevant error:

```bash
Collecting boto3==1.28.63 (from -r requirements.txt (line 6))
  Using cached boto3-1.28.63-py3-none-any.whl.metadata (6.7 kB)
ERROR: Cannot install botocore==1.31.63 and urllib3>=2.5.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested urllib3>=2.5.0
    botocore 1.31.63 depends on urllib3<2.1 and >=1.25.4; python_version >= "3.10"

To fix this you could try to:
1. Loosen the range of package versions you've specified
2. Remove package versions to allow pip to attempt to solve the dependency conflict

ERROR: ResolutionImpossible
```

Is there currently a recommended workaround for this?
Or do I need to wait until `botocore` is updated to support `urllib3>=2.5.0`?

Thanks in advance for your help!

**Comment by jmsmkn** [comment]
@plexoio you're trying to install `boto3==1.28.63` in your `requirements.txt (line 6)`. That version was released in October 2023. You need to update that version specification.
